o/assertstate: fix missing 'scheduled' header when auto refreshing assertions

MiguelPires opened the pull request:
When making requests to the snap store, some refresh requests with the 'fetch assertion' action were missing the "Snap-Refresh-Reason: scheduled" header. This was due to an unset RefreshOption when the assertions were auto refreshed. This PR sets the RefreshOptions when calling the store to make the refresh request.

https://bugs.launchpad.net/snapd/+bug/1937309

mardy commented on overlord/assertstate/bulk.go:
```diff
@@ -242,7 +242,7 @@ func (rpe *resolvePoolError) Error() string {
 }
 
-func resolvePool(s *state.State, pool *asserts.Pool, checkBeforeCommit func(*asserts.Database, asserts.Backstore) error, userID int, deviceCtx snapstate.DeviceContext) error {
+func resolvePool(s *state.State, pool *asserts.Pool, checkBeforeCommit func(*asserts.Database, asserts.Backstore) error, userID int, deviceCtx snapstate.DeviceContext, auto bool) error {
```
While in the other functions (which contain the word "refresh" in their name) it's quite obvious what the "auto" parameter does, here it isn't exactly so. Therefore I'd suggest renaming it to `autoRefresh`.

mardy replied:
Or `IsAutoRefresh`, I'm not sure which one is better :-)

mvo5 replied:
I will comment on this some more later (in the middle of meetings right now) but I am not a big fan ob non-obvious booleans in API. I.e. "foo(..., true)" will not tell much about what the bool means. I would suggest to follow the pattern we use in other function and pass a "resolveOptions struct" (name TBD) in that can be nil (and then the defaults will be used). On that struct we initially only have a single member "IsAutoRefresh" (or something like this). And then it becomes: "foo(..., &resolveOptions{IsAutoRefresh: true\)"

mardy commented on overlord/assertstate/bulk.go:
```diff
@@ -313,8 +313,8 @@ func resolvePool(s *state.State, pool *asserts.Pool, checkBeforeCommit func(*ass
 
-func resolvePoolNoFallback(s *state.State, pool *asserts.Pool, checkBeforeCommit func(*asserts.Database, asserts.Backstore) error, userID int, deviceCtx snapstate.DeviceContext) error {
-	err := resolvePool(s, pool, checkBeforeCommit, userID, deviceCtx)
+func resolvePoolNoFallback(s *state.State, pool *asserts.Pool, checkBeforeCommit func(*asserts.Database, asserts.Backstore) error, userID int, deviceCtx snapstate.DeviceContext, auto bool) error {
```
same here about the parameter name

mardy commented on overlord/assertstate/bulk.go:
```diff
@@ -175,7 +175,7 @@ func bulkRefreshValidationSetAsserts(s *state.State, vsets map[string]*Validatio
 	}
 
-	err := resolvePoolNoFallback(s, pool, beforeCommitChecker, userID, deviceCtx)
+	err := resolvePoolNoFallback(s, pool, beforeCommitChecker, userID, deviceCtx, true)
```
Using a variable would make it more clear what the last parameter is for:
```suggestion
	autoRefresh := true
	err := resolvePoolNoFallback(s, pool, beforeCommitChecker, userID, deviceCtx, autoRefresh)
```

mardy reviewed (commented):
A couple of minor comments :-)

But it doesn't seem to me that the commit message is relevant to these changes, as it talks about some header which I cannot see in the diff.

MiguelPires commented:
Thanks for the feedback @mardy :slightly_smiling_face:  I see your point, I'll amend the commit message to something like "Set missing RefreshOption when auto refreshing assertions" which is more relevant to the actual change

mvo5 commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ cmd/VERSION
 *~
 *.swp
 .vscode/
+.idea/
```
(nitpick^2) a bit unrelated to this PR but I understand that it's a bit silly to open a separate PR just for this change

mvo5 commented on overlord/assertstate/assertstate.go:
```diff
@@ -55,8 +55,18 @@ func findError(format string, ref *asserts.Ref, err error) error {
 
+// AutoRefreshSnapDeclarations refetches all the current snap declarations and their prerequisites while also setting
+// the store.RefreshOptions's IsAutoRefresh to reflect that it was automatic.
+func AutoRefreshSnapDeclarations(s *state.State, userID int) error {
```
I will leave this to @pedronis  to decide so maybe best wait for his opinion on this: I personally would slightly prefer to have an options struct on RefreshSnapDecleartions() instead of having two functions. Something like (names are strawmans right now):
```go
type RefreshSnapDeclarationsOptions struct {
    IsAutoRefresh bool
}
func RefreshSnapDeclarations(s *state.State, userID int, opts *RefreshSnapDeclarationsOptions) error {
    if opts == nil { opts = &RefreshSnapDeclarationsOptions{} };
...
```
and then pass this opts down into refreshSnapDeclarations, bulkRefreshSnapDeclarations etc.

But again, I will leave it to Samuele to make the final call on this. I think we want some kind of struct even if we keep two {Auto,}RefreshSnapDeclartions because bools in function declarations are a bit hard to read often. I.e. something like "SetAutomatic(true|false)" is easy to understand but for "refreshSnapDecarations(s, userID, false)" it's not obvious from reading the code in isolation what the "false" stands for and one needs to go back to the declaration of the method etc.

pedronis replied:
I think what was chosen, single function with options is fine. Pawel can tweak this further it later if needed

mvo5 commented on overlord/assertstate/assertstate_test.go:
```diff
@@ -962,6 +965,7 @@ func (s *assertMgrSuite) TestRefreshSnapDeclarationsNop(c *C) {
 
 	err := assertstate.RefreshSnapDeclarations(s.state, 0)
 	c.Assert(err, IsNil)
+	c.Assert(s.fakeStore.(*fakeStore).opts.IsAutoRefresh, Equals, false)
```
Thanks for adding this! 

As another nitpick, I think `c.Check()` is slightly better here and below but we have no hard rules for when to use Assert and Check, it's a bit mixed. Below there is a bunch more that the test could check even when the IsAutoRefresh is not set. In general the rule of thumb is that Assert() is useful if it does not make sense to test anything beyond the this point and Check when one detail about the test is not correct but it may make sense to continue. With assert() one has to fix issues one-by-one, i.e. run test, fix one assert, run again, fix next assert etc. With Check() multiple issues are visible in the same test run.

mvo5 reviewed (commented):
Thank you for this PR. I have a few inline comments, suggestions and nitpicks. But overall this looks very nice.

MiguelPires commented:
Thanks for the feedback @mvo5. I added an options struct to replace the less-readable bool in the API and changed the assertion to use Check.

mardy reviewed (approved):
Let's wait for one more approval, but this looks good to me. Thanks Miguel!

pedronis commented on overlord/assertstate/assertstate.go:
```diff
@@ -55,8 +55,16 @@ func findError(format string, ref *asserts.Ref, err error) error {
 	}
 }
 
+type RefreshDeclarationOptions struct {
```
this should follow the function name and have a plural:

RefreshSnapDeclarationsOptions

pedronis reviewed (approved):
lgtm, one comment about the options struct name

pedronis commented:
@mvo please re-review this as well

anonymouse64 commented on overlord/assertstate/assertstate.go:
```diff
@@ -336,7 +344,8 @@ func delayedCrossMgrInit() {
 // AutoRefreshAssertions tries to refresh all assertions
 func AutoRefreshAssertions(s *state.State, userID int) error {
-	if err := RefreshSnapDeclarations(s, userID); err != nil {
+	opts := &RefreshSnapDeclarationsOptions{IsAutoRefresh: true}
```
do we have a test case for checking that opts is true in this codepath? I see we do for ValidationSetAssertionForMonitor, but I'm unclear if we do for this case

MiguelPires replied:
Yes, I added an assertion to TestValidationSetAssertionsAutoRefresh. I can add two different tests to ensure that both RefreshSnapDeclarations and RefreshValidationSetAssertions propagate the options

anonymouse64 commented on overlord/assertstate/bulk.go:
```diff
@@ -175,7 +175,8 @@ func bulkRefreshValidationSetAsserts(s *state.State, vsets map[string]*Validatio
 	}
 
-	err := resolvePoolNoFallback(s, pool, beforeCommitChecker, userID, deviceCtx)
+	opts := &RefreshSnapDeclarationsOptions{IsAutoRefresh: true}
```
same question do we check that this option is specifically set to true in any test?

pedronis replied:
I misread this, this should come from outside through the levels for symmetry and also because some callers in the end will not be from auto-refresh I think

anonymouse64 reviewed (approved):
overall looks good, one nitpick about import ordering and questions about the test coverage, seems like we should try to have checks that the opts.IsAutoRefresh is true in more places, no?

pedronis commented on overlord/assertstate/assertstate.go:
```diff
@@ -336,7 +344,8 @@ func delayedCrossMgrInit() {
+	if err := RefreshSnapDeclarations(s, userID, opts); err != nil {
 		return err
 	}
 	return RefreshValidationSetAssertions(s, userID)
```
this should also take opts basically

MiguelPires replied:
Should I rename the options to RefreshAssertionsOptions since it's used in both kinds of assertions (snap declarations and validation sets)? Or is it better to have two different options for each type of call?

pedronis reviewed (changes requested):
I read things a bit too quickly it seems. The calls for validation-sets should also take options and respect them.

MiguelPires commented on daemon/export_test.go:
```diff
@@ -21,6 +21,7 @@ package daemon
 
 import (
 	"context"
+	"github.com/snapcore/snapd/overlord/assertstate"
```
Good to know, thanks! I'll re-order them

MiguelPires commented:
Thanks everyone for the feedback. I refactored the options so both the validation set and snap declaration assertions use it. I also added a few additional checks to some tests to insure both calls handle the opts correctly

pedronis commented:
@MiguelPires this needs a master merge as the now landed ValidationSetAssertionForEnforce needs changes too

pedronis reviewed (commented):
lgtm apart the need for a merge and fix

MiguelPires commented:
@stolowski I fixed the merge conflicts. Please have a look when you have a moment

MiguelPires commented:
@stolowski when you can, this PR is ready for a pass from you

pedronis reviewed (approved):
thanks

stolowski commented on overlord/assertstate/assertstate.go:
```diff
@@ -543,15 +557,19 @@ func ValidationSetAssertionForMonitor(st *state.State, accountID, name string, s
 	// refresh all currently tracked validation set assertions (this may or may not
 	// include the one requested by the caller).
-	if err = RefreshValidationSetAssertions(st, userID); err != nil {
+	if err = RefreshValidationSetAssertions(st, userID, opts); err != nil {
```
Shouldn't we simply pass have `opts := &RefreshAssertionsOptions{IsAutoRefresh: true}` here (and leave ValidationSetAssertionForEnforce signature intact - not take opts)?

MiguelPires replied:
The idea was just to make the API consistent. I think Samuele mentioned in an earlier comment that the functions should be consistent (but I may have misunderstood/misapplied the meaning)

stolowski replied:
Sorry, maybe I wasn't clear; I mean ValidationSetAssertionForMonitor doesn't take opts, but ValidationSetAssertionForEnforce does take them, so they are not consistent in that regard. Can we keep ValidationSetAssertionForEnfore consistent with  ValidationSetAssertionForMonitor and simply hardcode opts we pass to RefreshValidationSetAssertions here? I guess the question is when do we want to pass autorefresh:true and false here from the outside? On automatic snap refreshes vs manual installs maybe? CC @pedronis ?

pedronis replied:
sorry, I missed this skimming the last commit. We want:

```
opts := &RefreshAssertionsOptions{IsAutoRefresh: false}
```
(false!, not true)

here afaict.  We don't plan to use ValidationSetAssertionFor* in the context of auto refresh so they don't need to take options.

stolowski reviewed (changes requested):
Looks fine overall, needs one change discussed in the comments with Samuele.

stolowski reviewed (approved):
Thank you!

MiguelPires commented:
@mvo5 the tests failures are unrelated, can you merge this one as well please?
